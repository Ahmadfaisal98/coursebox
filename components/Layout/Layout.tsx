import { FC } from "react";
import Link from "next/link";
import styled from "@emotion/styled";

import { Logo } from "@/components/Logo";
import { Input } from "@/components/Input";
import { IconButton } from "@/components/IconButton";

const Wrapper = styled.div`
  display: grid;
  gap: 0.1rem;
  color: ${({ theme }) => theme.font.regular};
  background-color: ${({ theme }) => theme.background};
  padding: 0.5rem;
  grid-template-areas:
    "header"
    "nav"
    "search"
    "content"
    "footer";
  @media (min-width: 500px) {
    grid-template-columns: 1fr 3fr;
    grid-template-areas:
      "header  search"
      "nav     nav"
      "content content"
      "footer  footer";
    nav {
      flex-direction: row;
      justify-content: space-between;
    }
  }
  @media (min-width: 700px) {
    grid-template-columns: 1fr 4fr 2fr;
    grid-template-areas:
      "header  search  nav"
      "content content content"
      "footer  footer  footer";
    nav {
      flex-direction: row;
    }
  }
`;

const StyledLogo = styled(Logo)`
  grid-area: header;
  display: flex;
  align-items: center;
  height: 4rem;
  @media (max-width: 500px) {
    justify-content: center;
  }
`;

const LogoLink = styled.a`
  all: unset;
  cursor: pointer;
  &:hover {
    opacity: 0.9;
  }
`;

const MainNav = styled.nav`
  grid-area: nav;
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 0.5rem;
  a {
    cursor: pointer;
    color: ${({ theme }) => theme.font.regular};
    &:hover {
      opacity: 0.7;
    }
  }
`;

const SearchInput = styled(Input)`
  grid-area: search;
  width: 100%;
  height: 4rem;
`;

const Content = styled.main`
  grid-area: content;
`;

const Footer = styled.footer`
  grid-area: footer;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
  height: 5rem;
`;

export const Layout: FC = ({ children }) => (
  <Wrapper>
    <Link href="/" passHref>
      <LogoLink>
        <StyledLogo size={3}>C8X</StyledLogo>
      </LogoLink>
    </Link>
    <MainNav>
      <Link href="/all">All</Link>
      <Link href="/news">News</Link>
      <IconButton name="Moon" size={1} onClick={() => null} />
    </MainNav>
    <SearchInput icon="Search" placeholder="Search" onChange={() => null} />
    <Content>{children}</Content>
    <Footer>
      Â© {new Date().getFullYear()} NickOvchinnikov. All rights reserved.
    </Footer>
  </Wrapper>
);
