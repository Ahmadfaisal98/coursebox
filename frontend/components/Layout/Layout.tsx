import { useState, useLayoutEffect, useEffect, FC } from "react";
import Link from "next/link";
import { ThemeProvider } from "@emotion/react";

import { Themes } from "@/styles/themes";

import { IconButton } from "@/components/IconButton";
import { StyledLink } from "@/components/StyledLink";

import {
  Wrapper,
  LogoLink,
  StyledLogo,
  MainNav,
  SearchInput,
  Content,
  Footer,
} from "./components";

const useIsomorphicLayoutEffect =
  typeof window !== "undefined" ? useLayoutEffect : useEffect;

export const Layout: FC = ({ children }) => {
  const [isDark, setIsDark] = useState(true);

  const toggleDark = () => {
    localStorage.setItem("theme", isDark ? "light" : "dark");
    setIsDark(!isDark);
  };

  useIsomorphicLayoutEffect(() => {
    const theme = localStorage.getItem("theme");
    const themeExistsInStorage = Boolean(theme !== null);

    setIsDark(
      themeExistsInStorage
        ? Boolean(theme === "dark")
        : window.matchMedia("(prefers-color-scheme: dark)").matches
    );
  }, []);

  const theme = Themes[isDark ? "dark" : "light"];

  return (
    <ThemeProvider theme={theme}>
      <Wrapper>
        <Link href="/" passHref>
          <LogoLink>
            <StyledLogo size={3}>
              <span className="logo_short">C8X</span>
              <span className="logo_full">CoursesBox</span>
            </StyledLogo>
          </LogoLink>
        </Link>
        <MainNav>
          <Link href="/all" passHref>
            <StyledLink>All</StyledLink>
          </Link>
          <Link href="/login" passHref>
            <IconButton name="Login" size={1} />
          </Link>
          <IconButton
            name={!isDark ? "Moon" : "Sun"}
            size={1}
            onClick={toggleDark}
          />
        </MainNav>
        <SearchInput icon="Search" placeholder="Search" onChange={() => null} />
        <Content>{children}</Content>
        <Footer>
          Â© {new Date().getFullYear()} Ahmad Faisal. All rights reserved.
        </Footer>
      </Wrapper>
    </ThemeProvider>
  );
};
