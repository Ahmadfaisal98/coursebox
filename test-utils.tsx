import React, { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { Themes } from "./styles/themes";

const Wrapper: FC = ({ children }) => (
  <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
