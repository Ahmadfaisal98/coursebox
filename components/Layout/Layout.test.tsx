import { render } from "@/test-utils";

import { Layout } from "./Layout";

describe("Layout test cases", () => {
  const child = (
    <>
      <h1>Main article area</h1>
      <p>
        In this layout, we display the areas in source order for any screen less
        that 500 pixels wide. We go to a two column layout, and then to a three
        column layout by redefining the grid, and the placement of items on the
        grid.
      </p>
    </>
  );

  it("Render check", () => {
    const { asFragment } = render(<Layout>{child}</Layout>);

    expect(asFragment()).toMatchSnapshot();
  });
});
