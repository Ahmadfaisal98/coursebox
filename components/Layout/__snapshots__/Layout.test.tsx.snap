// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout test cases Render check 1`] = `
<DocumentFragment>
  <div
    class="css-9q9p2w"
  >
    <a
      class="css-1vuc8b5"
      href="/"
    >
      <header
        class="css-1he53g1"
        size="3"
      >
        C8X
      </header>
    </a>
    <nav
      class="css-gfwwxb"
    >
      <a
        href="/all"
      >
        All
      </a>
      <a
        href="/news"
      >
        News
      </a>
      <button
        class="css-1spd9kv"
        title="Moon"
      >
        <svg
          aria-label="Moon"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 24 24"
          width="1rem"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M10 6a8 8 0 0011.955 6.956C21.474 18.03 17.2 22 12 22 6.477 22 2 17.523 2 12c0-5.2 3.97-9.474 9.044-9.955A7.963 7.963 0 0010 6zm-6 6a8 8 0 008 8 8.006 8.006 0 006.957-4.045c-.316.03-.636.045-.957.045-5.523 0-10-4.477-10-10 0-.321.015-.64.045-.957A8.006 8.006 0 004 12zm14.164-9.709L19 2.5v1l-.836.209a2 2 0 00-1.455 1.455L16.5 6h-1l-.209-.836a2 2 0 00-1.455-1.455L13 3.5v-1l.836-.209A2 2 0 0015.29.836L15.5 0h1l.209.836a2 2 0 001.455 1.455zm5 5L24 7.5v1l-.836.209a2 2 0 00-1.455 1.455L21.5 11h-1l-.209-.836a2 2 0 00-1.455-1.455L18 8.5v-1l.836-.209a2 2 0 001.455-1.455L20.5 5h1l.209.836a2 2 0 001.455 1.455z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-1w3lqlr"
      height="7"
      width="20"
    >
      <div
        class="css-1v67xy3"
      >
        <input
          class="css-1bj11r8"
          placeholder="Search"
        />
        <svg
          aria-hidden="true"
          aria-label="Search"
          class="css-mjc9ur"
          data-icon="search"
          fill="currentColor"
          focusable="false"
          height="2rem"
          role="img"
          viewBox="64 64 896 896"
          width="2rem"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </div>
    </label>
    <main
      class="css-1qc1fvq"
    >
      <h1>
        Main article area
      </h1>
      <p>
        In this layout, we display the areas in source order for any screen less that 500 pixels wide. We go to a two column layout, and then to a three column layout by redefining the grid, and the placement of items on the grid.
      </p>
    </main>
    <footer
      class="css-1y0dtms"
    >
      Â© 2022 NickOvchinnikov. All rights reserved.
    </footer>
  </div>
</DocumentFragment>
`;
